### PR TITLE
refactor(assembly): add masl library serialization

### LINKED ISSUE #469: Serialization of Miden assembly libraries
Currently, we have only a single library - `stdlib`. When we serialize AST structure for the library, we put it directly into a Rust file (`asm.rs`). As we make the structure more general, it would probably be a good idea to have a stand-alone format which is language independent.

We can probably do something as simple as serialize ASTs (similar to how we do it now) into a separate binary file, and then refer to this file from `asm.rs`. The binary file should probably include the following data:

* Root namespace - e.g., for the standard library this would be `std`.
* Version - e.g., `v0.1`, `v0.2` etc.
* A map of serialized `ModuleAst` structs keyed by module path (relative to root namespace).

One open question: what should be the extension of this file. Two options come to mind:
1. `.mlib` - Miden Library.
2. `.mamb` - Miden Assembly Module Bundle (or Miden Assembly Module Binary).

I prefer the second option - but open to other suggestions.

### PR DESCRIPTION
**Pull request by vlopes11**
This commit introduces the serialization of library contents into a MASL (Miden Assembly Library) file. It will contain the root namespace, version, and a list of modules that will be available to the compilation.

This commit exposes the `Serializable` and `Deserializable` traits from the parsers module to be used as standard for the
serialization/deserialization strategies. It also simplifies the build process of the stdlib, creating a new instance on-demand via static bytes included during the build process.

closes #469

### CODE REVIEW

File: assembly/src/library.rs [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+/// Implementers of the library trait should use this base type to perform serialization into
+/// `masl` files.
+pub struct LibraryContents {
+    /// Root namespace of the library.
+    pub namespace: LibraryNamespace,
+    /// Version of the library.
+    pub version: Version,
+    /// Available modules.
+    pub modules: Vec<Module>,
+}
```
Comment (bobbinth) [inline assembly/src/library.rs:247]: nit: I don't love the name `LibraryContent`. Maybe it should be something like `DefaultLibrary` or something similar?

Also, I'm wondering if we should add another constructor here - something like `read_from_dir()`. This would read the library from a directory with `.masm` files - very similar to what happens in `build.rs` in `stdlib`. But this should probably done in another PR.

File: assembly/src/library.rs [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+mod use_std {
+    use super::*;
+    use std::{collections::HashSet, fs, io, path::Path};
+
+    impl LibraryContents {
```
Comment (bobbinth) [inline assembly/src/library.rs:350]: I was actually thinking of a slightly different interface here. Something more like:

```Rust
pub fn read_from_dir<P>(path: P) -> io::Result<Self>
pub fn write_to_file<P>(&self, path: P) -> io::Result<()>
```
The first one would read a library from a directory containing `.masm` files (basically, very similar to what is happening in `build.rs` of `stdlib`.

The second one would just write out the library as a single `.masl` file.

I do think that having a function for reading a set of libraries from a directory could be useful. But I'm not sure if there is a use case for writing a set of libraries to a directory. So, I'd probably keep the `read_dir()` (although maybe a different name is needed), but I would probably get rid of `write_to_dir()` with `write_to_file()`.

File: assembly/src/parsers/serde/deserialization.rs [hunk 3]
```diff
@@ -129,18 +129,41 @@ impl<'a> ByteReader<'a> {
 /// Returns `self` from its byte representation stored in provided `ByteReader` struct.
 pub trait Deserializable: Sized {
     fn read_from(bytes: &mut ByteReader) -> Result<Self, SerializationError>;
+
+    fn try_from_bytes(bytes: &[u8]) -> Result<Self, SerializationError> {
```
Comment (bobbinth) [inline assembly/src/parsers/serde/deserialization.rs:133]: nit: I would probably rename this to `read_from_bytes()` as the only thing different between this one and `read_from()` is this one takes byte slice as a parameter.

**Review by bobbinth — CHANGES_REQUESTED** [review]
This looks great! Thank you!

Not a full review yet, but I left some comments inline.

File: stdlib/src/lib.rs [hunk 4]
```diff
@@ -1,79 +1,50 @@
...
 /// TODO: add docs
 pub struct StdLibrary {
-    namespace: LibraryNamespace,
-    modules: Vec<Module>,
+    contents: LibraryContents,
+}
+
+impl Default for StdLibrary {
+    fn default() -> Self {
+        let bytes = include_bytes!("../assets/std.masl");
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply stdlib/src/lib.rs:15]: That is correct, we might have problems. I tested this a while ago for windows, but I just realized it was different circumstances.

I'm thinking of using `OUT_DIR`, and then copy it to `assets` so we always have it uploaded to git versioning. wdyt?

File: stdlib/build.rs [hunk 5]
```diff
@@ -13,69 +10,57 @@ use md_renderer::MarkdownRenderer;
...
-/// file under `./src/asm.rs`.
-///
-/// The `asm.rs` file exports a single static array of string tuples. Each tuple consist of module
-/// namespace label and the module source code for this label.
+/// Read and parse the contents from `./asm` into a `LibraryContents` struct, serializing it into
+/// `assets` folder under `std` namespace.
 #[cfg(not(feature = "docs-rs"))]
 fn main() -> io::Result<()> {
     // re-build the `./src/asm.rs` file only if something in the `./asm` directory has changed
     println!("cargo:rerun-if-changed=asm");
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply stdlib/build.rs:39]: True, fixed!

File: assembly/src/library.rs [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+            patch,
+        })
+    }
+}
+
+// LIBRARY
+// ================================================================================================
+
+/// A library definition that provides AST modules for the compilation process.
+pub trait Library {
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply assembly/src/library.rs:219]: Fixed!

File: assembly/src/library.rs [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+        L: Library,
+        for<'a> <L as Library>::ModuleIterator<'a>: ExactSizeIterator,
+    {
+        Self::write_contents_into(
+            library.root_ns(),
+            library.version(),
+            library.modules(),
+            target,
+        )
+    }
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply assembly/src/library.rs:313]: That is true, I added that as a possible helper for the user, but we don't really need it for now. Removed for the sake of simplicity!

File: assembly/src/library.rs (hunk 1 above)
Comment (vlopes11) [reply assembly/src/library.rs:247]: The implementation attempts to solve the issue of now knowing the available namespaces in the given directory. This way, it will load all `masl` files in a given dir, and the user can just pick a namespace from there.

If we specify a file, that will work as well, but when the user will have to know the full path

File: assembly/src/parsers/serde/deserialization.rs (hunk 3 above)
Comment (vlopes11) [reply assembly/src/parsers/serde/deserialization.rs:133]: Fixed!

File: stdlib/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bobbinth) [reply stdlib/src/lib.rs:15]: I think we should just test on a windows machine. I can do this later today.

File: assembly/src/lib.rs [hunk 8]
```diff
@@ -14,18 +14,24 @@ use vm_core::{
...
 mod procedures;
 use procedures::{CallSet, Procedure};
 pub use procedures::{ProcedureId, ProcedureName};
 
 mod parsers;
-pub use parsers::{parse_module, parse_program, ModuleAst, ProcedureAst, ProgramAst};
+pub use parsers::{
+    parse_module, parse_program, ByteReader, ByteWriter, Deserializable, ModuleAst, ProcedureAst,
+    ProgramAst, Serializable,
+};
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply assembly/src/lib.rs:28]: btw, I'm now thinking that maybe we should move out these structs as well as `Serializable` and `Deserializable` traits out of parsers as there are quite a few changes there.

File: stdlib/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (vlopes11) [reply stdlib/src/lib.rs:16]: I added a OS check. We can also run the CI tests on a windows machine if we include this in the CI. wdyt?

```
  tests-win:
    needs: tests
    runs-on: [windows-2022]
    strategy:
      matrix:
        rust:
          - nightly
    steps:
      - uses: actions/checkout@v2
      - uses: actions-rs/toolchain@v1
        with:
          profile: minimal
          toolchain: ${{ matrix.rust }}
          override: true
      - uses: actions-rs/cargo@v1
        with:
          command: test
          args: --all-features
```

File: assembly/src/library.rs (hunk 1 above)
Comment (vlopes11) [reply assembly/src/library.rs]: That is correct. This validation is an attempt to simplify future compilation libraries in a directory. If we follow that, we can just read all directory contents and load these modules, and we are guaranteed to not have duplicates.

It is ofc optional and we can remove this check.

File: assembly/src/library.rs (hunk 2 above)
Comment (vlopes11) [reply assembly/src/library.rs]: Refactored with a small change on the output function: we just take a dir, and we create a masl with the namespace as file name.

We can quickly revert this change if it is not desirable.

File: stdlib/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bobbinth) [reply stdlib/src/lib.rs:16]: Let's do it!

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks good! Thank you! I left a few more small comments inline.

File: assembly/src/library.rs [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+            Ok(Self {
+                namespace,
+                version,
+                modules,
+            })
+        }
+
+        /// Write the library to a target dir, using its namespace as file name and the appropriate
+        /// extension.
+        pub fn write_to_dir<P>(&self, path: P) -> io::Result<()>
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply assembly/src/library.rs:173]: Actually, I take the above back. I think `write_to_dir()` is fine. The only potential change is that maybe we should rename `path` parameter into `dir_path`?

File: assembly/src/library.rs [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+pub struct BasicLibrary {
+    /// Root namespace of the library.
+    pub namespace: LibraryNamespace,
+    /// Version of the library.
+    pub version: Version,
+    /// Available modules.
+    pub modules: Vec<Module>,
+}
+
+impl Library for BasicLibrary {
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply assembly/src/library.rs:42]: Since we serialize this type as `masl`, I think its best if we go for `MaslLibrary`

File: assembly/src/library.rs [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+    fn default() -> Self {
+        Self::MIN
+    }
+}
+
+impl Serializable for Version {
+    fn write_into(&self, target: &mut ByteWriter) -> Result<(), SerializationError> {
+        target.write_u64(self.major);
+        target.write_u64(self.minor);
+        target.write_u64(self.patch);
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply assembly/src/library.rs:380]: Its just to be compatible with the broadly used `semver`

https://github.com/dtolnay/semver/blob/05316ff842e0cab71587294de894e4f822932a14/src/lib.rs#L161-L168

But in fact we don't really need `u64` there

File: assembly/src/library.rs [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+    }
+
+    // VALIDATIONS
+    // --------------------------------------------------------------------------------------------
+
+    /// Validate if the module belongs to the provided namespace.
+    pub fn check_namespace(&self, namespace: &LibraryNamespace) -> Result<(), LibraryError> {
+        (self.path.namespace() == namespace.as_str())
+            .then_some(())
+            .ok_or(LibraryError::EmptyProcedureName)
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply assembly/src/library.rs:260]: True, fixed!

File: assembly/src/library.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (vlopes11) [reply assembly/src/library.rs:173]: True, fixed!

File: assembly/src/library.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (vlopes11) [reply assembly/src/library.rs]: Thinking again, it's definitely unrealistic to have 24 bytes to represent the version. If we ever need to convert to the `semver` crate, we just cast the `u16` and we will be fine.

Fixed!

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!
